Ensure lifecycle tasks wait for messages to be pushed

francoisferrand opened the pull request:
lifecycle task pushes new entries to bucket topic, but may commit before the entry is commited : which allows multiple lifeycle iterations to happen in parallel.

Issue: BB-641

francoisferrand committed:
Ensure lifecycle tasks wait for messages to be pushed

Issue: BB-641

bert-e commented:
# Hello francoisferrand,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/backbeat/doc/user).


<details>
  <summary><b>Available options</b></summary>

  name   | description  | privileged   | authored
  ------ | ------------ | ------------ |----------
  `/after_pull_request` | Wait for the given pull request id to be merged before continuing with the current one. |  | 
  `/bypass_author_approval` | Bypass the pull request author's approval |  :star:  | 
  `/bypass_build_status` | Bypass the build and test status |  :star:  | 
  `/bypass_commit_size` | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 
  `/bypass_incompatible_branch` | Bypass the check on the source branch prefix |  :star:  | 
  `/bypass_jira_check` | Bypass the Jira issue check |  :star:  | 
  `/bypass_peer_approval` | Bypass the pull request peers' approval |  :star:  | 
  `/bypass_leader_approval` | Bypass the pull request leaders' approval |  :star:  | 
  `/approve` | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 
  `/create_pull_requests` | Allow the creation of integration pull requests. |  | 
  `/create_integration_branches` | Allow the creation of integration branches. |  | 
  `/no_octopus` | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
  `/unanimity` | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
  `/wait` | Instruct Bert-E not to run until further notice. |  | 
  

</details>



<details>
  <summary><b>Available commands</b></summary>

  name   | description  | privileged
  ------ | ------------ | ------------
  `/help` | Print Bert-E's manual in the pull request. | 
  `/status` | Print Bert-E's current status in the pull request ```TBA``` | 
  `/clear` | Remove all comments from Bert-E from the history ```TBA``` | 
  `/retry` | Re-start a fresh build ```TBA``` | 
  `/build` | Re-start a fresh build ```TBA``` | 
  `/force_reset` | Delete integration branches & pull requests, and restart merge process from the beginning. | 
  `/reset` | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
  
</details>




*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-641 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.6.55`

* `9.0.0`


Please check the `Fix Version/s` of BB-641, or the target
branch of this pull request.

bert-e commented:
# Request integration branches



Waiting for integration branch creation to be requested by the user.

To request integration branches, please comment on this pull request with the following command:

```
/create_integration_branches
```

Alternatively, the `/approve` and `/create_pull_requests` commands will automatically
create the integration branches.

williamlardier commented on extensions/lifecycle/tasks/LifecycleTaskV2.js:
```diff
@@ -236,8 +252,10 @@ class LifecycleTaskV2 extends LifecycleTask {
+                    return resolve();
+                })));
+
+            return Promise.allSettled(promises).then(() => done(), done);
```
With this code, in case of error in `_compareRulesToList`, we are ignoring the returned error, as we would have it in the "then" with the `rejected` status. This looks like a behavior change we don't want? We ignore for `_sendBucketEntry` but for `_compareRulesToList`, we actually call the `reject` function...

`Promise.allSettled` will return an array with the status for each promise, including the ones rejected:

```
Promise.allSettled([
  Promise.resolve(33),
  new Promise((resolve) => setTimeout(() => resolve(66), 0)),
  99,
  Promise.reject(new Error("an error")),
]).then((values) => console.log(values), err => console.log('catch', err));

[
    {
        "status": "fulfilled",
        "value": 33
    },
    {
        "status": "fulfilled",
        "value": 66
    },
    {
        "status": "fulfilled",
        "value": 99
    },
    {
        "status": "rejected",
        "reason": {}
    }
]
```

(and no "catch" log).

francoisferrand replied:
you are right (and I will update the code), but actually `_compareRulesToList` is not upposed to fail either : it calls `_retryEntry`, which ignores the errors. Errors on individual objects are retried a bit ; but eventually we just skip the object (or batch), relying on the periodic scan to eventually process these obejcts.

francoisferrand committed:
Make allSettled usage safer

allSettled does not follow the usuage fullfil pattern: it will never
reject, and always fullfil with an array of the results of each
promises.

This is not an issue in the case of lifecycle, where we actually ignore
all errors; but it makes the code look inconsistent, as it suggests
errors are possible but not handle them.

To avoid future issues, add proper processing of the results of
allSettled to build a single error when appropriate.

Issue: BB-641

williamlardier commented on extensions/lifecycle/tasks/LifecycleTaskV2.js:
```diff
@@ -255,7 +257,9 @@ class LifecycleTaskV2 extends LifecycleTask {
 
-            return Promise.allSettled(promises).then(() => done(), done);
+            return Promise.allSettled(promises).then(results => done(
+                results.findLast(r => r.status === 'rejected')[0]?.reason
```
Shouldn't it be

```suggestion
                results.findLast(r => r.status === 'rejected')?.reason
```

as it an array of dimension 1, and findLast only returns the first matching element?

francoisferrand replied:
yes, switch from filter to findLast on the last minute...

nicolas2bert reviewed (changes requested):
Waiting for the Kafka message delivery report callback may not be ideal for at least two reasons:

1. The message is already considered as locally consumed even before it reached the queue processor queue.
The lifecycle conductor relies on `KafkaBacklogMetrics` to calculate the lag. 
`KafkaBacklogMetrics` uses `consumer.position()` (for consumers) and `queryWatermarkOffsets()` (for producers/clients) to fetch offset info, then writes those offsets into Zookeeper.
`consumer.position()` in node-rdkafka returns the consumer’s current read position for each assigned partition, i.e. the local offset that the consumer is about to read next (offset of the last consumed message + 1). It does not return the “committed offset” that Kafka itself stores for the consumer group.  A message is considered “locally consumed” as soon as the consumer (via librdkafka under the hood) has handed that message to `BackbeatConsumer`. Basically, right after consumer.consume(...) is called and the message is shown in your callback or event handler, the offset for that message is advanced in the local consumer state.

2. It will slow down our process since `sendToTopic` callback is called when Kafka returns the delivery report for this message. Delivery reports are invoked asynchronously after the producer has handed off messages to Kafka. Ifwe decide to block or wait synchronously for that delivery report every time we send a message, it will impact performance and throughput.

francoisferrand commented:
> The message is already considered as locally consumed even before it reached the queue processor queue

That is a fair point (and may actually help on another issue), but I don't really see how this is a problem _for this change_: handling an entry by the bucket processor typically takes at least one second already (scanning & checking the state of every object), so we face this discrepancy anyway...

This change is simply about ensuring that the we keep the "slot" until the entry is "fully" processed, instead of leaving many things pending: which can be an issue esp. since we are listing pushing continuation messages.

What am I missing here?

> Ifwe decide to block or wait synchronously for that delivery report every time we send a message, it will impact performance and throughput.

In theory yes;
Practically however, since we are processing up to 1000 entries at a time, I wonder if this makes a real impact: most of the reports would be received in the time we process each entry... (except for very small buckets, in which case throughput may not be so important)

It is certainly a trade off, but consistent processing seems important as well: or do you think it is completely safe to leave all these messages dangling, and already start processing next message(s)?

nicolas2bert commented:
> What am I missing here?

My understanding was that the goal of this PR is to prevent multiple lifecycle iterations (triggered by Conductor) from running in parallel. I just pointed out that the lag is based on the “locally consumed” offset rather than on a processed or stored offset. So even if we wait for an entry to be fully processed, it won't stop the bucket-lifecycle topic lag from being zero while there are still other bucket messages in the pipeline.

> most of the reports would be received in the time we process each entry

Regarding the internal lifecycle listing, it does not necessarily return a 1000 objects; it only includes those that meet the specified criteria (prefix, age, etc...) from the next 10,000 entries. We might even end up with a listing response containing only a few objects, or none at all. 
NOTE: This 10,000 entry limit helps avoid placing excessive load on Metadata by preventing the evaluation of an unbounded number of entries.

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-641 contains:


* `8.6.56`

* `9.0.19`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.6.57`

* `9.0.19`

* `9.1.1`


Please check the `Fix Version/s` of BB-641, or the target
branch of this pull request.

bert-e commented:
# Jira issue not found



The Jira issue BB-641 was not found.
